### DISCUSSION TITLE
BUG: Block drag and drop interaction fails within vscode extension webview

### DISCUSSION THREAD

**Discussion by protozoo** [GrapesJS/grapesjs#4253] [Ideas]
### GrapesJS version

- [X] I confirm to use the latest version of GrapesJS

### What browser are you using?

VSCode (WebView within VSCode extension)

### Reproducible demo link

https://grapesjs.com/demo.html

### Describe the bug

**How to reproduce the bug?**
1. Create a new vscode extension (or use one that you already have)
2. Create a Webview that contains an HTML document with an iframe loading the GrapesJS demo (see example code below)
3. Run the extension and webview to load the GrapesJS demo
4. Try to drag a Grapes Block in the canvas

**What is the expected behavior?**
The dragged block should be added to the canvas

**What is the current behavior?**
The dragged component is NOT added to the canvas. 
You can watch a video showing the problem here: https://www.protozoo.com/tests/grapesjs/vscode_extension_grapes.mov. As demonstrated in the video, other dragging interactions work as expected. The issue can be seen near the end.

If is necessary to execute some code in order to reproduce the bug, paste it here below:
```js
// Here's a simplified code example on how to create a webview within a vscode example, and load grapesjs demo inside.
// You can download a ready-to-build-and-run project here: 
// http://www.protozoo.com/tests/grapesjs/visualEditor_vscode_extension.zip

const vscode = require('vscode');

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
	console.log('Congratulations, your extension "visualEditor" is now active!');
	let disposable = vscode.commands.registerCommand('visualeditor.openVisualEditor', function () {
		const panel = vscode.window.createWebviewPanel(
			'visualEditor',
			'Visual Editor',
			vscode.ViewColumn.One,
			{	enableScripts: true	}
		  );
	
		  
		  panel.webview.html = `<!DOCTYPE html>
		  <html lang="en">
		  <head>
			  <meta charset="UTF-8">
			  <meta name="viewport" content="width=device-width, initial-scale=1.0">
			  <title>Visual Editor</title>
		  </head>
		  <body style="padding: 0 0px;">
			  <iframe src="https://grapesjs.com/demo.html" style="position: fixed; width: 100%; height: 100%;">
			</iframe>
		  </body>
		  </html>`
	});

	context.subscriptions.push(disposable);
}

function deactivate() {}

module.exports = {
	activate,
	deactivate
}

```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by artf** [comment]
Hi @protozoo I'm not sure exactly why but looks like the vscode iframe (probably sandboxed) is blocking the native HTML5 Drag & Drop (this is what is used on Blocks).
As this is not strictly related to the core I'll move this one to the discussion.

**Reply by protozoo** [reply]
Hi @artf many thanks for the response. One additional thing I found yesterday is that if I remove the "draggable" property to blocks, then they work as expected (well, not really since the visual representation of the dragged element is not displayed while you drag, but at least the "drag" event and callback gets executed, the canvas shows the green line-pointer indicating where the component will be dropped, and the drop works).

So, by now I just set "draggable" to false everywhere (I edited [this line](https://github.com/artf/grapesjs/blob/dc0be299e0471b8e34dd0aae02c9c61a1e61c695/src/block_manager/view/BlockView.js#L135) so that it's always "false" ), but I assume this is probably going to cause all other sorts of problems.

Any suggestion on how to fix the problem better than this patch?
Thanks again!

**Reply by artf** [reply]
Yeah by removing draggable from blocks you basically disable the native HTML5 D&D and let the previous implementation do the work, but as I didn't work on it for a long time, I'd not expect indeed to see it work properly.
The proper way to fix that is to understand why the native D&D is disabled and if there is an option to enable it.

**Reply by protozoo** [reply]
All right, I'll report this to VS Code mantainers, I already opened an issue there. I guess this info will help. Thanks!

**Reply by protozoo** [reply]
Hi @artf . I never got a useful response to [my issue at vscode's repo](https://github.com/microsoft/vscode/issues/147148) but the workaround I mentioned above (disabling the "draggable" attribute) proved good enough until now. 

However I am now needing to customize the block manager quite a lot (not only look and feel, but also behavior), so I created my own block manager component in Vue, following the [instructions at Grapesjs docs](https://grapesjs.com/docs/modules/Blocks.html#customization).

So far so good, the blocks are listed properly and work as expected... in the browser. But, once again, inside vscode (as a webview) drag and dropping blocks to the canvas will not work. I tried adding the attribute draggable=false but it didn't work this time. I've been trying to debug it and looks like the drag events that the default block manager assigns to the block elements are not being assigned to the block elements of the custom block manager.

I see this by opening the devtools, selecting one of the block elements, and checking the Event Listeners tab.
The block elements of the default manager list many of them, including drag, dragend, dragenter, dragstart....

If I switch to the custom manager, most of these are missing.

Any idea how I could fix this?

Thanks a lot

**Reply by artf** [reply]
If you've only implemented the native HTML5 D&D in your custom block manager, removing the draggable won't dispatch any event (the default block manager has a fallback). You should probably implement your own D&D implementation (eg. using pointer events) so it can also work in the vscode iframe.

**Reply by protozoo** [reply]
Thanks for the response @artf. Please could you point to where/how you do that in the default block manager so I can try to do the same or similar? I've been trying to find it without success. Thanks

**Reply by artf** [reply]
You can find it here: https://github.com/artf/grapesjs/blob/9a3a5f06d315430f108f74773e87d357183603e1/src/block_manager/view/BlockView.ts#L23-L24

**Reply by protozoo** [reply]
Checking it out, many thanks
